### PR TITLE
Adding digitization for Hyper-K OD

### PR DESCRIPTION
**Pull request by P3tru**
The OD digitization has been added when using the HyperKOD geometry.

Many different classes had to be modified to being able to create a new instance of hits and digi collection to be separated from the ID output tree. Now the output of the simulation is written into 2 different WCSimRootEvent, the first one is left unchanged and the second one have the suffix "_OD", and contain an identical structure to easily analyze the OD information by using scripts similar to the ones already provided.

The previous PR with the geometry already stored the OD part of the HitsCollection inside a different structure. This PR add the digitization of these HitsCollection at the end of the EventAction. Messenger classes have also been modified to change the dark rates independantly  between the ID and OD PMTs. There is no separate trigger at the moment, only the ID part, but the OD can be studied using the "NoTrigger" algorithm implenented recently.

Physics validation plots are being produced and will be released shortly.

### CODE REVIEW

File: src/WCSimWCTrigger.cc [hunk 1]
```diff
@@ -595,3 +609,29 @@ void WCSimWCTriggerNDigits2::DoTheWork(WCSimWCDigitsCollection* WCDCPMT) {
   bool ndigits_test = true;
   AlgNDigits(WCDCPMTCopy, remove_hits, ndigits_test);
 }
+
+// *******************************************
+// DERIVED CLASS
+// *******************************************
+
+
+WCSimWCTriggerNoTrigger::WCSimWCTriggerNoTrigger(G4String name,
```
Comment (eosullivan) [inline src/WCSimWCTrigger.cc]: I think this is from another pull request which has already been merged (#220). Or is there some change here that goes with this PR? If not, please update so this change is removed.

**Comment by eosullivan** [comment]
I notice that the OD PMT info isn't being stored in the wcsimGeoT tree. Also, you should include some examples in the sample-root-scripts directory of how to access the OD information.

**Comment by eosullivan** [comment]
I am also wondering how flexible this new digitization scheme is. So, for example, there is some interest to see what the simulation looks like if you include different types of PMTs in the ID. These would then need to be digitized separately. Can this scheme be easily adapted for more general cases than just 1 ID PMT type + 1 OD PMT type?

**Comment by eosullivan** [comment]
Have you done any initial validation on this PR yet? If you are able to come to the software call this week, it would be nice if you could show some validation plots so we can talk it over.

File: src/WCSimWCTrigger.cc (hunk 1 above)
Comment (P3tru) [reply src/WCSimWCTrigger.cc]: Yes I had to add inside the mother class the detector element to instantiate separate trigger algorithm, I have also cleaned the file.

**Comment by P3tru** [comment]
I notice that the OD PMT info isn't being stored in the wcsimGeoT tree. Also, you should include some examples in the sample-root-scripts directory of how to access the OD information. : 
 - OD PMTs info are constructed inside geometry tables classes and dumped in the geofile.txt. I have never used the wcsimGeoT tree, let me know if you want me to add the OD information inside. I guess this is not related to the OD digitization, I could create a separate PR to do that.

I am also wondering how flexible this new digitization scheme is. So, for example, there is some interest to see what the simulation looks like if you include different types of PMTs in the ID. These would then need to be digitized separately. Can this scheme be easily adapted for more general cases than just 1 ID PMT type + 1 OD PMT type?
 - Yes so now you can instantiate any DAQ and trigger classes with a name, contained inside the detectorElement member. Proceeding the same way, one could just add a new WC collection and instantiate a new digitization operating on this collection.

**Comment by P3tru** [comment]
Here comes the slides with the validation of the pull request : 
[HyperKOD.pdf](https://github.com/WCSim/WCSim/files/1759329/HyperKOD.pdf)

**Comment by eosullivan** [comment]
@P3tru, the merge of #228 had some overlap with your pull request, and so there are now some conflicts here. Can you look this over and resolve the conflicts? It looks like they are contained to WCSimEventAction.cc.

**Comment by P3tru** [comment]
Just merged everything and running very quick check on SuperK geometry to see if everything still consistent. I really have no time at the moment to make more checks, everything seems fine and don't really understand why they were conflicts in the first place --- I really made no modification on the conflicted hunks.

**Comment by eosullivan** [comment]
Could you update this branch once more (sorry for the repeated requests)? Ben has merged in some changes that allow the full validation to be online now, and I think this pull request could be a good test to see it in action.

**Comment by P3tru** [comment]
Ok just done it.
After a few checks it seems that Federico's PR acts normal on the output file for the develop branch but there is seems to be some conflicts with the OD filling tree function. I am investigating that right now and try to fix it asap...

**Comment by brichards64** [comment]
Hi Stephane,

Did you change the the ouput root file format?

**Comment by P3tru** [comment]
Hi Ben, 

I added an other tree with the OD info, but its structure is duplicated from the normal one. There is now 2 wcsimrootsuperevent inside the output tree for the ID and OD part.
It seems that since Federico's PR and mine are incompatible at the moment and I am trying to fix it. But my macros are working for both trees, so what are the problem that you're seeing ?

**Comment by brichards64** [comment]
Hi Stephane,

not sure as yet your code compiles and creates a root file as far as i can tell but then the script the Tom provided for gernerating the reference plots seems to be segging. Ill download your branch and see if i can unpick the issue.

**Comment by brichards64** [comment]
OK crash is caused by the TTree wcsimT haveing the wrong number of entries.

I generated 10 Events so expected 10 Entries and actually there were quoted as 8018. When i tried to access entry 10, it segfaulted (naturally as that many entries dont exist.   

Long64_t nevent = tree->GetEntries(); is how the entries are found and i checked the root file output and indeed there apear to be 8018 entries in the wcsimT tree.

Which means wither Stephane you have changed something to massivly increase the number of fills that ocurr on the wcsimT, however these fills arnt valid entries (or not in the expected format). Or somthing in your code change is corrupting the entries variable in the wcsimT ttree.

**Comment by brichards64** [comment]
The plot thinkens......

So your code the execution time is orders of magnitude longer than the current head (which kind of marries up with the extra fills 

Here is the current head output from the wcsimT tree.

` root [1] wcsimT.Print()
******************************************************************************
*Tree    :wcsimT    : WCSim Tree                                             *
*Entries :       10 : Total =         3176608 bytes  File  Size =    1024008 *
*        :          : Tree compression factor =   3.10                       *
******************************************************************************
*Branch  :wcsimrootevent                                                     *
*Entries :       10 : BranchElement (see below)                              *
*............................................................................*
*Br    0 :fUniqueID : UInt_t                                                 *
*Entries :       10 : Total  Size=       1334 bytes  File Size  =        820 *
*Baskets :       10 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*
*Br    1 :fBits     : UInt_t                                                 *
*Entries :       10 : Total  Size=       1398 bytes  File Size  =        900 *
*Baskets :       10 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*
*Br    2 :fEventList : TObjArray*                                            *
*Entries :       10 : Total  Size=    3172296 bytes  File Size  =    1021302 *
*Baskets :       10 : Basket Size=      64000 bytes  Compression=   3.11     *
*............................................................................*`


and your pull request looks like this:

`root [7] wcsimT->Print()
******************************************************************************
*Tree    :wcsimT    : WCSim Tree                                             *
*Entries :     8018 : Total =         4999006 bytes  File  Size =    1423289 *
*        :          : Tree compression factor =   3.51                       *
******************************************************************************
*Branch  :wcsimrootevent                                                     *
*Entries :       10 : BranchElement (see below)                              *
*............................................................................*
*Br    0 :fUniqueID : UInt_t                                                 *
*Entries :       10 : Total  Size=       1334 bytes  File Size  =        820 *
*Baskets :       10 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*
*Br    1 :fBits     : UInt_t                                                 *
*Entries :       10 : Total  Size=       1398 bytes  File Size  =        900 *
*Baskets :       10 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*
*Br    2 :fEventList : TObjArray*                                            *
*Entries :       10 : Total  Size=    4991982 bytes  File Size  =    1420300 *
*Baskets :       10 : Basket Size=      64000 bytes  Compression=   3.51     *
*............................................................................*
*Branch  :wcsimrootevent_OD                                                  *
*Entries :        0 : BranchElement (see below)                              *
*............................................................................*
*Br    3 :fUniqueID : UInt_t                                                 *
*Entries :        0 : Total  Size=        494 bytes  One basket in memory    *
*Baskets :        0 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*
*Br    4 :fBits     : UInt_t                                                 *
*Entries :        0 : Total  Size=        478 bytes  One basket in memory    *
*Baskets :        0 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*
*Br    5 :fEventList : TObjArray*                                            *
*Entries :        0 : Total  Size=        498 bytes  One basket in memory    *
*Baskets :        0 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*`

**Comment by P3tru** [comment]
Hi Ben, sorry I fixed the bug a bit earlier it was a bad set of the nb of tree event inside WCSimEventAction, I am readapting my OD macro and will push everything by tonight, thanks for investigating !

**Comment by brichards64** [comment]
Fun and interesting results from the validation this time: https://wcsim.github.io/Validation/

Overall Failed:   (mac files can be found here for reference https://github.com/WCSim/Validation/tree/master/Generate/macReference)
1) code fully compiled
2) code fully executed
3) So the execution test (generating 10 500MeV electrons) failed. output was made and looking at the validation plots the distributions of:
```
ntracks Error!!!
nrawhits Error!!!
ntubeshitraw Error!!!
ntracks Error!!!
nrawhits Error!!!
ntubeshitraw Error!!!
digipeperdigi Error!!!
digitimeperdigi Error!!!
totaldigipe Error!!!
totaldiginoisefrac Error!!!
```
all no have differnt shapes.

4) However interestingly the 10Mev 100Mev and 500Mev electron tests all passed (there maybe a difference in the mac files how these particles are generated from that of the execution test).
5) all the muon events Failed: These however failed to make validation plots. So as pefore something in the root output file must be seg faulting the print out script (possibly the same issue possibly a new one with the output format being screwed up)
6) the Pion events look to have produced correctly. However notably the time for execution has  gone up by 30%. This longer execution time is not seen in any of the electron tests.

So yeh fun to unpick all that :).

**Comment by brichards64** [comment]
I also note that the 800 MeV muons seemed to finisk their simulation in only 6 mins rather than the standard ~30. which is also interesting. 

all consol output can be found here https://travis-ci.org/WCSim/WCSim/builds/349448208?utm_source=github_status&utm_medium=notification

**Comment by brichards64** [comment]
So did a quick test of one of the muon runs to see why the segfaulted. The 200MeV muon run i reduced to only 10 events and again the wcsimT tree entry size is too big.

```
root [1] wcsimT->Print()
******************************************************************************
*Tree    :wcsimT    : WCSim Tree                                             *
*Entries :       15 : Total =         3046825 bytes  File  Size =     688373 *
*        :          : Tree compression factor =   4.43                       *
******************************************************************************
*Branch  :wcsimrootevent                                                     *
*Entries :       10 : BranchElement (see below)                              *
*............................................................................*
*Br    0 :fUniqueID : UInt_t                                                 *
*Entries :       10 : Total  Size=       1334 bytes  File Size  =        820 *
*Baskets :       10 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*
*Br    1 :fBits     : UInt_t                                                 *
*Entries :       10 : Total  Size=       1398 bytes  File Size  =        900 *
*Baskets :       10 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*
*Br    2 :fEventList : TObjArray*                                            *
*Entries :       10 : Total  Size=    3039801 bytes  File Size  =     685397 *
*Baskets :       10 : Basket Size=      64000 bytes  Compression=   4.43     *
*............................................................................*
*Branch  :wcsimrootevent_OD                                                  *
*Entries :        0 : BranchElement (see below)                              *
*............................................................................*
*Br    3 :fUniqueID : UInt_t                                                 *
*Entries :        0 : Total  Size=        494 bytes  One basket in memory    *
*Baskets :        0 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*
*Br    4 :fBits     : UInt_t                                                 *
*Entries :        0 : Total  Size=        478 bytes  One basket in memory    *
*Baskets :        0 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*
*Br    5 :fEventList : TObjArray*                                            *
*Entries :        0 : Total  Size=        498 bytes  One basket in memory    *
*Baskets :        0 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*
```

Not as much as before but looks like you still ahve an issue there and that's the cause of the muon problems. Interesting that id doesnt effect the electrons tho.

**Comment by brichards64** [comment]
As expected the 800mu crash was cuased by too many events also.

```
root [1] wcsimT->Print()
******************************************************************************
*Tree    :wcsimT    : WCSim Tree                                             *
*Entries :       17 : Total =         4969975 bytes  File  Size =    1161950 *
*        :          : Tree compression factor =   4.28                       *
******************************************************************************
*Branch  :wcsimrootevent                                                     *
*Entries :       10 : BranchElement (see below)                              *
*............................................................................*
*Br    0 :fUniqueID : UInt_t                                                 *
*Entries :       10 : Total  Size=       1334 bytes  File Size  =        820 *
*Baskets :       10 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*
*Br    1 :fBits     : UInt_t                                                 *
*Entries :       10 : Total  Size=       1398 bytes  File Size  =        900 *
*Baskets :       10 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*
*Br    2 :fEventList : TObjArray*                                            *
*Entries :       10 : Total  Size=    4962951 bytes  File Size  =    1158956 *
*Baskets :       10 : Basket Size=      64000 bytes  Compression=   4.28     *
*............................................................................*
*Branch  :wcsimrootevent_OD                                                  *
*Entries :        0 : BranchElement (see below)                              *
*............................................................................*
*Br    3 :fUniqueID : UInt_t                                                 *
*Entries :        0 : Total  Size=        494 bytes  One basket in memory    *
*Baskets :        0 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*
*Br    4 :fBits     : UInt_t                                                 *
*Entries :        0 : Total  Size=        478 bytes  One basket in memory    *
*Baskets :        0 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*
*Br    5 :fEventList : TObjArray*                                            *
*Entries :        0 : Total  Size=        498 bytes  One basket in memory    *
*Baskets :        0 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*
```

**Comment by brichards64** [comment]
So more bad news, turns out that the Execution test has more errors that reported, as i forgot to turn on the Error text for if the numbers of entries wernt matching. You can clearly see on a number of plots they dont match. Iv corrected it tho so any future tests will show this.

It is very interesting that the electron tests all passed but the Execution test didnt. both simulate electrons so there must be a differnce in the mac file.

**Comment by eosullivan** [comment]
Thanks @P3tru for being the validation testing guinea pig. 

What is the exact difference between the execution test and the electron tests? Is the energy different?

**Comment by brichards64** [comment]
Having never relly used WCSim im unsure but i can show you whats in the mac files and maybe we can work it out. Tthere are three energies in the electron tests 10MeV 100MeV and 500MeV exch produce 10 events. The Execution test produces 10 events of 500 MeV electrons.

**Execution test:**
```
/WCSim/PMTQEMethod     Stacking_Only
/WCSim/PMTCollEff on
/WCSim/SavePi0 false
/DAQ/Digitizer SKI
/DAQ/Trigger NDigits
/control/execute macros/daq.mac
/DarkRate/SetDarkMode 1
/DarkRate/SetDarkHigh 100000
/DarkRate/SetDarkLow 0
/DarkRate/SetDarkWindow 4000
/mygen/generator gun
/gun/particle e-
/gun/energy 500 MeV
/gun/direction 1 0 0
/gun/position 0 0 0
/WCSimIO/RootFile Execution_Test.root
/run/beamOn 10
```

**500 MeV electrons:**


```
/run/verbose 0
/tracking/verbose 0
/hits/verbose 0
/WCSim/PMTQEMethod     Stacking_Only
/WCSim/SavePi0 false
/mygen/generator gun
/gun/particle e-
/gun/energy 500 MeV
/gun/direction 1 0 0
/gun/position 0 0 0
/WCSimIO/RootFile 500e.root
/WCSimIO/Timer true
/run/beamOn 100
```

**Comment by brichards64** [comment]
seems to me mainly darkrate and DAQ/digitiser and DAQ/trigger

**Comment by eosullivan** [comment]
When I check out this pull request branch and run the execution test macro from the console, it seems to run a bit slow, but the number of events looks right. So, something seems weird here...

root [1] wcsimT->Print()
******************************************************************************
*Tree    :wcsimT    : WCSim Tree                                             *
*Entries :       10 : Total =         5032366 bytes  File  Size =    1445359 *
*        :          : Tree compression factor =   3.48                       *
******************************************************************************
*Branch  :wcsimrootevent                                                     *
*Entries :       10 : BranchElement (see below)                              *
*............................................................................*
*Br    0 :fUniqueID : UInt_t                                                 *
*Entries :       10 : Total  Size=       1334 bytes  File Size  =        820 *
*Baskets :       10 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*
*Br    1 :fBits     : UInt_t                                                 *
*Entries :       10 : Total  Size=       1398 bytes  File Size  =        900 *
*Baskets :       10 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*
*Br    2 :fEventList : TObjArray*                                            *
*Entries :       10 : Total  Size=    5025342 bytes  File Size  =    1442377 *
*Baskets :       10 : Basket Size=      64000 bytes  Compression=   3.48     *
*............................................................................*
*Branch  :wcsimrootevent_OD                                                  *
*Entries :        0 : BranchElement (see below)                              *
*............................................................................*
*Br    3 :fUniqueID : UInt_t                                                 *
*Entries :        0 : Total  Size=        494 bytes  One basket in memory    *
*Baskets :        0 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*
*Br    4 :fBits     : UInt_t                                                 *
*Entries :        0 : Total  Size=        478 bytes  One basket in memory    *
*Baskets :        0 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*
*Br    5 :fEventList : TObjArray*                                            *
*Entries :        0 : Total  Size=        498 bytes  One basket in memory    *
*Baskets :        0 : Basket Size=      64000 bytes  Compression=   1.00     *
*............................................................................*

**Comment by brichards64** [comment]
No that's the same as what I said. The execution test has the correct number of events. Its the Muon ones that dont. The execution test is the one that gives slightly differnt distributions.

**Comment by brichards64** [comment]
Thats an improvement. now all of the output files were readable.

But still differences in the output

**Comment by brichards64** [comment]
Hi all,

ok something is wrong, i was seeing before that i had to run the validation multiple times to get all green and now the merge just failed. So we still ahve an intermittent issue. This may be hard to find, but it may also be that updating the other mac files in a similar way makes it go away.

However i cant see why the same issue would cause an intermittent error..... so it may be something new.....

I could however try to update the mac files and then run it a few more times to see if the intermittent error disappears.

what do people think?

thanks

Ben



On Thu, 11 Oct 2018 at 20:01, eosullivan <notifications@github.com<mailto:notifications@github.com>> wrote:

Merged #229<https://github.com/WCSim/WCSim/pull/229> into develop.

**Comment by eosullivan** [comment]
That's confusing. You mean that if you run the validation again, with an identical pull request and validation macros, that it might pass?

Let's try an propagate this dark rate fix through the macros and see what happens.

**Pull request merged**